### PR TITLE
Fix ALE sponge diagnostics

### PR DESCRIPTION
**Pull request by WenhaoChen89**
This PR is to fix the ALE sponge diagnostics. 

In the module MOM_ALE_sponge, the ALE sponge diagnostics of tracers is initialized with a fixed order, i.e., the first one is 'temperature' and the second one is 'salinity' (line 604-609 in subroutine init_ALE_sponge_diags). However, when the data is posted for the diagnostics, it follows the order of how tracer sponges are set up (line 961-997 in subroutine apply_ALE_sponge).

The fix includes:
1. The set_up_ALE_sponge_field is modified to include the tracer details diagnostics registration, namely name, long name, and unit.
2. The type p2d is changed to include three extra fields, name, long name, and unit.
3. The init_ALE_sponge_diags is modified to initialize the ALE sponges of tracers in the order of how set_up_ALE_sponge_field is called.

### CODE REVIEW

**Comment by marshallward** [comment]
I don't see much structure to these commits, so I will do a squash-merge.

**Review by marshallward — COMMENTED** [review]
A minor style change removing spaces around `=` in optional. function arguments

**Comment by marshallward** [comment]
@WenhaoChen89 I made a minor change to this PR fixing a style issue (removal of spaces around operators inside function arguments).

(I admit that I also just wanted to try the in-browser editor and see if it would work :sweat_smile:)

**Comment by marshallward** [comment]
Gaea regression: https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/16465 :heavy_check_mark:

File: src/initialization/MOM_state_initialization.F90 [hunk 1]
```diff
@@ -2135,9 +2135,11 @@ subroutine initialize_sponges_file(G, GV, US, use_temperature, tv, u, v, depth_t
       if (use_temperature) then
         allocate(tmp_tr(isd:ied,jsd:jed,nz_data))
         call MOM_read_data(filename, potemp_var, tmp_tr(:,:,:), G%Domain, scale=US%degC_to_C)
-        call set_up_ALE_sponge_field(tmp_tr, G, GV, tv%T, ALE_CSp)
+        call set_up_ALE_sponge_field(tmp_tr, G, GV, tv%S, ALE_CSp, 'temp', &
```
Comment (marshallward) [inline src/initialization/MOM_state_initialization.F90:2138]: Should this be `tv%T`?

File: src/parameterizations/vertical/MOM_ALE_sponge.F90 [hunk 2]
```diff
@@ -81,6 +81,9 @@ module MOM_ALE_sponge
   real :: scale = 1.0  !< A multiplicative factor by which to rescale input data
   real, dimension(:,:), pointer :: p => NULL() !< pointer the data.
   real, dimension(:,:), pointer :: h => NULL() !< pointer the data grid.
+  Character (:), allocatable  :: name  !< The name of the input field
+  Character (:), allocatable  :: long_name !< The long name of the input field
+  Character (:), allocatable  :: unit !< The unit of the input field
```
Comment (marshallward) [inline src/parameterizations/vertical/MOM_ALE_sponge.F90:86]: Use `character(:), allocatable`.  No caps or space before `(:)`

File: src/parameterizations/vertical/MOM_ALE_sponge.F90 [hunk 3]
```diff
@@ -134,7 +137,7 @@ module MOM_ALE_sponge
   logical :: tripolar_N !< grid is folded at its north edge
 
   !>@{ Diagnostic IDs
-  integer, dimension(2) :: id_sp_tendency      !< Diagnostic ids for temperature and salinity
+  integer, dimension(MAX_FIELDS_) :: id_sp_tendency      !< Diagnostic ids for temperature and salinity
```
Comment (marshallward) [inline src/parameterizations/vertical/MOM_ALE_sponge.F90:140]: Comment probably needs to be updated

File: src/user/DOME2d_initialization.F90 [hunk 4]
```diff
@@ -473,12 +473,10 @@ subroutine DOME2d_initialize_sponges(G, GV, US, tv, depth_tot, param_file, use_A
       enddo
     enddo ; enddo
 
-    if ( associated(tv%T) ) then
-      call set_up_ALE_sponge_field(T, G, GV, tv%T, ACSp)
-    endif
-    if ( associated(tv%S) ) then
-      call set_up_ALE_sponge_field(S, G, GV, tv%S, ACSp)
-    endif
+    if ( associated(tv%T) ) call set_up_ALE_sponge_field(T, G, GV, tv%S, ACSp, 'temp', &
```
Comment (marshallward) [inline src/user/DOME2d_initialization.F90:476]: Same here, should `tv%S` be `tv%T`

File: src/user/ISOMIP_initialization.F90 [hunk 5]
```diff
@@ -626,12 +626,11 @@ subroutine ISOMIP_initialize_sponges(G, GV, US, tv, depth_tot, PF, use_ALE, CSp,
     ! momentum is typically not damped within the sponge.                !
 
     !  The remaining calls to set_up_sponge_field can be in any order. !
-    if ( associated(tv%T) ) then
-      call set_up_ALE_sponge_field(T, G, GV, tv%T, ACSp)
-    endif
-    if ( associated(tv%S) ) then
-      call set_up_ALE_sponge_field(S, G, GV, tv%S, ACSp)
-    endif
+    if ( associated(tv%T) ) call set_up_ALE_sponge_field(T, G, GV, tv%S, ACSp, 'temp', &
```
Comment (marshallward) [inline src/user/ISOMIP_initialization.F90:629]: `tv%S` -> `tv%T`

File: src/user/RGC_initialization.F90 [hunk 6]
```diff
@@ -168,8 +168,10 @@ subroutine RGC_initialize_sponges(G, GV, US, tv, u, v, depth_tot, PF, use_ALE, C
     call initialize_ALE_sponge(Idamp, G, GV, PF, ACSp, h, nz)
 
     !  The remaining calls to set_up_sponge_field can be in any order.
-    if ( associated(tv%T) ) call set_up_ALE_sponge_field(T, G, GV, tv%T, ACSp)
-    if ( associated(tv%S) ) call set_up_ALE_sponge_field(S, G, GV, tv%S, ACSp)
+    if ( associated(tv%T) ) call set_up_ALE_sponge_field(T, G, GV, tv%S, ACSp, 'temp', &
```
Comment (marshallward) [inline src/user/RGC_initialization.F90:171]: `tv%S` -> `tv%T`

File: src/user/dense_water_initialization.F90 [hunk 7]
```diff
@@ -279,8 +279,10 @@ subroutine dense_water_initialize_sponges(G, GV, US, tv, depth_tot, param_file,
       enddo
     enddo
 
-    if (associated(tv%T)) call set_up_ALE_sponge_field(T, G, GV, tv%T, ACSp)
-    if (associated(tv%S)) call set_up_ALE_sponge_field(S, G, GV, tv%S, ACSp)
+    if ( associated(tv%T) ) call set_up_ALE_sponge_field(T, G, GV, tv%S, ACSp, 'temp', &
```
Comment (marshallward) [inline src/user/dense_water_initialization.F90:282]: `tv%S` -> `tv%T`

Comment (WenhaoChen89) [reply src/user/dense_water_initialization.F90]: Thanks Marshall. I have corrected all these.

**Review by marshallward — APPROVED** [review]

**Comment by marshallward** [comment]
Gaea regression (v2): https://gitlab.gfdl.noaa.gov/ogrp/MOM6/-/pipelines/16476 :heavy_check_mark:

**Pull request merged**